src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -175,22 +206,115 @@ private ITextBuffer TryGetTextBufferFromDocData(object docData)
+        /// Otherwise, we schedule a task on foreground task scheduler.
         /// </summary>
         public void NotifyDocumentRegisteredToProject(IVisualStudioHostDocument document)
         {
+            if (IsForeground())
+            {
+                NotifyDocumentRegisteredToProject_Core(document, cancellationToken: CancellationToken.None);
+            }
+            else
+            {
+                var cts = new CancellationTokenSource();
+                var task = InvokeBelowInputPriority(() => NotifyDocumentRegisteredToProject_Core(document, cts.Token), cts.Token);
+                AddPendingDocumentInitializationTask(document, task, cts);
+            }
+        }
+
+        private void AddPendingDocumentInitializationTask(IVisualStudioHostDocument document, Task task, CancellationTokenSource cts)
+        {
+            var taskAndTokenSource = new TaskAndTokenSource() { Task = task, CancellationTokenSource = cts };
+            lock (_gate)
+            {
+                // Add taskAndTokenSource to the pending document initialization tasks.
+                // Check for cancellation before adding as the task might already have been completed/cancelled/faulted before we reached here.
+                if (!cts.IsCancellationRequested && !task.IsCompleted && !task.IsFaulted)
+                {
+                    _pendingDocumentInitializationTasks.Add(document.Id, taskAndTokenSource);
+                }
+            }
+        }
+
+        private void CancelPendingDocumentInitializationTask(IVisualStudioHostDocument document)
+        {
+            lock (_gate)
+            {
+                CancelPendingDocumentInitializationTask_NoLock(document);
+            }
+        }
+
+        private void CancelPendingDocumentInitializationTasks_NoLock(IEnumerable<DocumentKey> documentKeys)
+        {
+            foreach (var documentKey in documentKeys)
+            {
+                StandardTextDocument document;
+                if (_documentMap.TryGetValue(documentKey, out document))
+                {
+                    CancelPendingDocumentInitializationTask_NoLock(document);
+                }
+            }
+        }
+
+        private void CancelPendingDocumentInitializationTask_NoLock(IVisualStudioHostDocument document)
+        {
+            // Cancel pending initialization task for the document, if any, and dispose the cancellation token source.
+            TaskAndTokenSource taskAndTokenSource;
+            if (_pendingDocumentInitializationTasks.TryGetValue(document.Id, out taskAndTokenSource))
+            {
+                // Signal cancellation and wait for task to complete.
+                taskAndTokenSource.CancellationTokenSource.Cancel();
+                taskAndTokenSource.Task.Wait();
+                _pendingDocumentInitializationTasks.Remove(document.Id);
+            }
+        }
+
+        private void RemovePendingDocumentInitializationTask(IVisualStudioHostDocument document)
```

Comment:
Yes, the difference was just the Task.Wait() in the cancel implementation above. Given we don't need it, both of this can be the same method.